help_function/src/onedpl_test_temporary_allocation.cpp

```diff
@@ -0,0 +1,156 @@
+// ====------ onedpl_test_temporary_allocation.cpp---------- -*- C++ -*
+// ----===////
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//
+// ===----------------------------------------------------------------------===//
+
+#include "oneapi/dpl/algorithm"
+#include "oneapi/dpl/execution"
+#include "oneapi/dpl/iterator"
+
+#include "dpct/dpct.hpp"
+#include "dpct/dpl_utils.hpp"
+
+#include <iostream>
+
+#include <sycl/sycl.hpp>
+
+#include <sys/resource.h>
+
+template <typename String, typename _T1, typename _T2>
+int ASSERT_EQUAL(String msg, _T1 &&X, _T2 &&Y) {
+  if (X != Y) {
+    std::cout << "FAIL: " << msg << " - (" << X << "," << Y << ")" << std::endl;
+    return 1;
+  }
+  return 0;
+}
+
+int test_passed(int failing_elems, std::string test_name) {
+  if (failing_elems == 0) {
+    std::cout << "PASS: " << test_name << std::endl;
+    return 0;
+  }
+  return 1;
+}
+
+int main() {
+  // used to detect failures
+  int failed_tests = 0;
+  int num_failing = 0;
+
+  // Test One, temporary allocation with device tag.
+  {
+    std::size_t num_elements = 10;
+    dpct::device_sys_tag device_sys;
+    auto [int64_ptr, elements_allocated] =
+        dpct::get_temporary_allocation<int64_t>(device_sys, num_elements);
+    int64_t num_data[10];
+    std::iota(std::rbegin(num_data), std::rend(num_data), 1);
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(int64_ptr, &num_data, sizeof(int64_t) * num_elements);
+    });
+    dpct::get_default_queue().wait();
+
+    dpct::get_default_queue().submit([&](sycl::handler &h) {
+      h.memcpy(&num_data, int64_ptr, num_elements * sizeof(int64_t));
+    });
+    dpct::get_default_queue().wait();
+
+    std::string test_name =
+        "get and release temporary allocation - device memory with tag";
+    failed_tests += ASSERT_EQUAL(test_name, elements_allocated, num_elements);
+    for (std::size_t i = 0; i != num_elements; ++i)
+      failed_tests += ASSERT_EQUAL(test_name, num_data[i],
+                                   static_cast<int64_t>(num_elements - i));
+
```

Comment:
Same comment as the other file, need to make sure the copy works in both directions by checking the device memory and or manipulating the data in device memory before returning it.

I wont tag each instance, but all device tag versions should have a similar update.

Reply:
Done